### PR TITLE
Allow limiting New-PI credit to partner institutions

### LINKED ISSUE #94: Make sure Credit Code 0002 is only for MGHPCC Partners
## Motivation
Ensure the 0002 credit code $1000 is only for institutions with an MGHPCC partnership.

## Completion Criteria
Update invoice script to only include the institutions with a MGHPCC partnership.

## Description
- [ ] Update script to only include institutions that are MGHPCC affiliates, it would be worth adding this list as a table that we can reference.
   - BU
   - HU
   - MIT
   - Northeastern
   - Umass
   - Yale

## Completion dates
Desired - 2024-10-02
Required - TBD

### PR DESCRIPTION
**Pull request by QuanMPhm**
Closes #94, `process_report.py` will now use `nerc_rates` to determine if the New-PI credit is limited to institutions that are MGHPCC partners for the given invoice month.

The script determines whether an institution's MGHPCC partnership is active by first seeing if the `mghpcc_partnership_start_date` field is set in `institute_list.yaml`, then checks if the start date is within or before the invoice month.

@joachimweyl To mark an institution as an MGHPCC partner, you will need to set the field `mghpcc_partnership_start_date` in `institute_list.yaml` to the start date of the partnership in YYYY-MM format.

I.e for Boston University, their current entry looks like this:
```
-   display_name: Boston University
    domains:
    - bu.edu
    - robbaron
```

To mark them as a partner, change their entry like so:
```
-   display_name: Boston University
    domains:
    - bu.edu
    - robbaron
    mghpcc_partnership_start_date: "2024-10"
```

### CODE REVIEW

File: process_report/institute_list.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+-   display_name: Northeastern University
+    domains:
+    - northeastern.edu
+-   display_name: Boston University
+    domains:
+    - bu.edu
+    - robbaron
```
Comment (knikolla) [inline process_report/institute_list.yaml:7]: Since we've fully transitioned to using email addresses rather than username (and also have a aliasin system in place), I feel it's safe to remove everything that is an username from this file.

**Review by knikolla — COMMENTED** [review]
Not a full review yet, just one comment related to the data in the file.

File: process_report/invoices/billable_invoice.py [hunk 2]
```diff
@@ -115,6 +116,31 @@ def export_s3(self, s3_bucket):
...
 
+    def _filter_partners(self, data):
+        active_partnerships = list()
+        institute_list = util.load_institute_list()
+        for institute_info in institute_list:
+            if partnership_start_date := institute_info.get(
+                "mghpcc_partnership_start_date", None
+            ):
+                if (
+                    util.get_month_diff(self.invoice_month, partnership_start_date[:-3])
```
Comment (naved001) [inline process_report/invoices/billable_invoice.py:127]: Usually I'd add a comment when you need to access some very specific indexes of a string. 

In this case, however, I think we should convert it to a datetime object and then get the month and year. 
```
>>> "{:%Y-%m}".format(datetime.strptime("2024-09-01", "%Y-%m-%d"))
'2024-09'
```
This would also work in the case where someone writes the date as "2024-9-1" in institute_list.yaml because nothing is enforcing that a zero is added.

```
>>> "{:%Y-%m}".format(datetime.strptime("2024-9-1", "%Y-%m-%d"))
'2024-09'
```
as opposed to

```
>>> "2024-9-1"[:-3]
'2024-'
```

File: process_report/util.py [hunk 3]
```diff
@@ -27,21 +27,33 @@ def get_invoice_bucket():
...
-    institution_name = institute_map.get(institution_key, "")
+    institution_domain = pi_uname.split("@")[-1]
+    for i in range(institution_domain.count(".") + 1):
+        if institution_name := institute_map.get(institution_domain, ""):
+            break
+        institution_domain = institution_domain[institution_domain.find(".") + 1 :]
 
     if institution_name == "":
-        logger.warn(f"PI name {pi_uname} does not match any institution!")
+        print(f"Warning: PI name {pi_uname} does not match any institution!")
```
Comment (naved001) [inline process_report/util.py:47]: why was this changed from logging to printing?

**Review by naved001 — COMMENTED** [review]
Looks great!, just some questions and comments. Thanks @QuanMPhm !

File: process_report/invoices/billable_invoice.py [hunk 4]
```diff
@@ -140,19 +166,19 @@ def get_initial_credit_amount(
         )
         print(f"New PI Credit set at {new_pi_credit_amount} for {self.invoice_month}")
 
-        current_pi_set = set(data[invoice.PI_FIELD])
+        credit_eligible_projects = self._get_credit_eligible_projects(data)
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply process_report/invoices/billable_invoice.py:169]: @naved001 I think I'll make an issue in the near future to replace every `print()` statement with logging. @knikolla @naved001 I will use your judgement on when that "near future" is.

File: process_report/util.py (hunk 3 above)
Comment (QuanMPhm) [reply process_report/util.py:47]: Ah, must have been a mistake when I was moving some code around

File: process_report/institute_list.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    - yale.edu
+-   display_name: Dartmouth College
+    domains:
+    - dartmouth.edu
+-   display_name: Photrek
+    domains:
+    - kenric.nelson@photrek.io
+-   display_name: Positron Networks
+    domains:
+    - srao@positronnetworks.com
```
Comment (knikolla) [inline process_report/institute_list.yaml:64]: why do you include the entire email here rather than just the domain? It seems this wouldn't work, since you're splitting at the `@` when processing a PI's email.

File: process_report/invoices/billable_invoice.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (knikolla) [reply process_report/invoices/billable_invoice.py:170]: I don't think it's necessary to log this for now, since the credit eligible projects will have a credit code associated to them (unless they have no credit).

**Review by knikolla — CHANGES_REQUESTED** [review]

File: process_report/institute_list.yaml (hunk 5 above)
Comment (joachimweyl) [reply process_report/institute_list.yaml:64]: that was my fault in this [PR](https://github.com/CCI-MOC/process_csv_report/pull/90).

Comment (QuanMPhm) [reply process_report/institute_list.yaml:64]: This was also my fault for not catching it during review. I made a fix in #98

File: process_report/invoices/billable_invoice.py [hunk 6]
```diff
@@ -115,6 +117,31 @@ def export_s3(self, s3_bucket):
...
 
+    def _filter_partners(self, data):
+        active_partnerships = list()
+        institute_list = util.load_institute_list()
+        for institute_info in institute_list:
+            if partnership_start_date := institute_info.get(
+                "mghpcc_partnership_start_date"
+            ):
+                partnership_year_month = "{:%Y-%m}".format(
+                    datetime.strptime(partnership_start_date, "%Y-%m-%d")
```
[the remark this answers is not in the document]
Comment (naved001) [reply process_report/invoices/billable_invoice.py:128]: @knikolla I think that was because of my concerns described in this comment: https://github.com/CCI-MOC/process_csv_report/pull/96#discussion_r1761622405

[the remark this answers is not in the document]
Comment (naved001) [reply process_report/invoices/billable_invoice.py:128]: but yeah, if it's just year and month then it could probably be passed directly.

[the remark this answers is not in the document]
Comment (QuanMPhm) [reply process_report/invoices/billable_invoice.py]: The date format of the partnership start date will now be YYYY-MM. I've updated the PR message accordingly.

**Review by naved001 — APPROVED** [review]
looks fine to me.


> There is also some small cleanup. Namely, some functions in `process_report.py`
that have been moved to `util.py` were not removed during refactoring.

What was the small cleanup? whatever was moved from process_report.py to util.py in this PR seemed to be relevant?

**Comment by naved001** [comment]
> While writing this PR, I just decided the two functions should be placed in utils.py,

In the future I'd recommend separate commits for tangentially related cleanups.
